### PR TITLE
unary op for const int

### PR DESCRIPTION
**Pull request by yumuzi234**

### CODE REVIEW

File: pl/sempass/const_op.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import (
+	"shanhu.io/smlvm/lexing"
+	"shanhu.io/smlvm/pl/tast"
+	"shanhu.io/smlvm/pl/types"
+)
+
+func unaryOpConst(b *builder, opTok *lexing.Token, B tast.Expr) tast.Expr {
+	op := opTok.Lit
+	bref := B.R()
+	if !bref.IsSingle() {
```
Comment (yumuzi234) [inline pl/sempass/const_op.go]: probably this error will never be triggered

File: pl/sempass/const_op.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			"float is not supported yet")
+		return nil
+	}
+	switch op {
+	case "+":
+		return B // just shortcut this
+	case "-":
+		return tast.NewConst(tast.NewRef(types.NewConstInt(-v, code)))
+	}
+	b.CodeErrorf(opTok.Pos, "pl.invalidOp",
```
Comment (yumuzi234) [inline pl/sempass/const_op.go]: now we do not support ^, then either int or unit is OK here

File: pl/sempass/int_op.go [hunk 3]
```diff
@@ -6,100 +6,6 @@ import (
 	"shanhu.io/smlvm/pl/types"
 )
 
-func unaryOpConst(b *builder, opTok *lexing.Token, B tast.Expr) tast.Expr {
```
Comment (yumuzi234) [inline pl/sempass/int_op.go]: separate const_op from int_op

File: pl/sempass/switch.go [hunk 4]
```diff
@@ -100,6 +100,9 @@ func buildCaseExpr(b *builder, m map[int64][]ast.Expr,
 	}
 	if !types.IsConst(t) {
 		e = constCast(b, pos, v, e, t)
+	} else {
```
Comment (yumuzi234) [inline pl/sempass/switch.go]: t is the type of the expr of the switch, now we confine that expr for switch can only be integer and expr for case can only be const, so types.IsConst(t) is always false for now.

File: pl/sempass/operand.go [hunk 5]
```diff
@@ -90,7 +90,14 @@ func buildIdent(b *builder, ident *lexing.Token) tast.Expr {
 	case tast.SymVar, tast.SymField:
 		ref := tast.NewAddressableRef(t)
 		return &tast.Ident{Token: ident, Ref: ref, Sym: s}
-	case tast.SymConst, tast.SymStruct, tast.SymType, tast.SymImport:
+	case tast.SymConst:
+		if types.IsConst(t) {
+			ref := tast.NewRef(t)
```
Comment (h8liu) [inline pl/sempass/operand.go]: you can just inline this to the next line.

File: pl/sempass/switch.go [hunk 6]
```diff
@@ -100,6 +100,9 @@ func buildCaseExpr(b *builder, m map[int64][]ast.Expr,
...
+		// TODO const switch expr
```
Comment (h8liu) [inline pl/sempass/switch.go]: TODO: (add a colon)

File: pl/types/const.go [hunk 7]
```diff
@@ -10,14 +10,24 @@ type Const struct {
...
+// NewConstInt creates a new constant for a specific int type.
+func NewConstInt(v int64, t T) *Const {
+	if !IsInteger(t) {
+		panic("the type for NewConstInt must be a int or unit")
+	}
+	return &Const{Value: v, Type: t}
+}
 
 // NewConstString creates a new string constant.
 func NewConstString(s string) *Const {
```
Comment (h8liu) [inline pl/types/const.go]: you said you will remove this?

File: pl/types/const.go [hunk 8]
```diff
@@ -49,7 +59,17 @@ func IsConst(t T) bool {
 	return ok
 }
 
+// ConstType checks and transforms a type to const type.
+func ConstType(t T) (*Const, bool) {
```
Comment (h8liu) [inline pl/types/const.go]: remove this function; this has almost no different with simply t.(*Const).

File: pl/sempass/const_op.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+		return nil
+	}
+
+	ct, ok := types.ConstType(bref.T)
+	if !ok {
+		b.CodeErrorf(opTok.Pos, "pl.expectConstExpr",
+			"expect const expression but got %q %s", op, B)
+		return nil
+	}
+	code, ok := ct.Type.(types.Basic)
```
Comment (h8liu) [inline pl/sempass/const_op.go]: this should be `t` rather than `code`. It is not a really code (outside the types package).

Comment (h8liu) [inline pl/sempass/const_op.go]: you can just use `types.IsBasic(types.Bool)`, `types.IsInteger()`

File: pl/sempass/const_op.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+		return B // just shortcut this
+	case "-":
+		return tast.NewConst(tast.NewRef(types.NewConstInt(-v, code)))
+	}
+	b.CodeErrorf(opTok.Pos, "pl.invalidOp",
+		"invalid operation on int const: %q on %s", op, B)
+	return nil
+}
+
+func binaryOpConst(b *builder, opTok *lexing.Token, A, B tast.Expr) tast.Expr {
```
Comment (h8liu) [inline pl/sempass/const_op.go]: I assume you have not changed anything in this function yet.

File: pl/sempass/operand.go (hunk 5 above)
Comment (yumuzi234) [inline pl/sempass/operand.go]: what does this mean?

File: pl/types/const.go (hunk 7 above)
Comment (yumuzi234) [inline pl/types/const.go]: forgot

File: pl/types/const.go (hunk 8 above)
Comment (yumuzi234) [inline pl/types/const.go]: OK

File: pl/sempass/const_op.go (hunk 9 above)
Comment (yumuzi234) [inline pl/sempass/const_op.go]: I was thinking about to distinguish uint and int at the beginning, so want to keep what kind of basic type it is. You are right now IsInteger and IsBasic will do the job

File: pl/sempass/const_op.go (hunk 10 above)
Comment (yumuzi234) [inline pl/sempass/const_op.go]: not yet

File: pl/bare_func_test.go [hunk 11]
```diff
@@ -223,6 +223,11 @@ func TestBareFunc_bad(t *testing.T) {
 	o("invalidExprStmt", "_:= nil+nil")
 	o("invalidExprStmt", "true + 3") // boolean cannot add
```
Comment (h8liu) [inline pl/bare_func_test.go]: could you add some test to track what you implemented in this pr?

**Pull request merged**